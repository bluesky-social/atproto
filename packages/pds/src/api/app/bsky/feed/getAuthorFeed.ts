import { Server } from '../../../../lexicon'
import AppContext from '../../../../context'
import { authPassthru } from '../../../proxy'
import { OutputSchema } from '../../../../lexicon/types/app/bsky/feed/getAuthorFeed'
import { isReasonRepost } from '../../../../lexicon/types/app/bsky/feed/defs'
import {
  LocalViewer,
  handleReadAfterWrite,
  LocalRecords,
} from '../../../../read-after-write'

export default function (server: Server, ctx: AppContext) {
  server.app.bsky.feed.getAuthorFeed({
    auth: ctx.authVerifier.accessOrRole,
    handler: async ({ req, params, auth }) => {
      const requester =
        auth.credentials.type === 'access' ? auth.credentials.did : null
      const res = await ctx.appViewAgent.api.app.bsky.feed.getAuthorFeed(
        params,
        requester ? await ctx.serviceAuthHeaders(requester) : authPassthru(req),
      )
      if (requester) {
        return await handleReadAfterWrite(ctx, requester, res, getAuthorMunge)
      }
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })
}

const getAuthorMunge = async (
  localViewer: LocalViewer,
  original: OutputSchema,
  local: LocalRecords,
  requester: string,
): Promise<OutputSchema> => {
  const localProf = local.profile
  // only munge on own feed
  if (!isUsersFeed(original, requester)) {
    return original
  }
  let feed = original.feed
  // first update any out of date profile pictures in feed
  if (localProf) {
    feed = feed.map((item) => {
      if (item.post.author.did === requester) {
        return {
          ...item,
          post: {
            ...item.post,
            author: localViewer.updateProfileViewBasic(
              item.post.author,
              localProf.record,
            ),
          },
        }
      } else {
        return item
      }
    })
  }
  feed = await localViewer.formatAndInsertPostsInFeed(feed, local.posts)
  return {
    ...original,
    feed,
  }
}

const isUsersFeed = (feed: OutputSchema, requester: string) => {
  const first = feed.feed.at(0)
  if (!first) return false
  if (!first.reason && first.post.author.did === requester) return true
  if (isReasonRepost(first.reason) && first.reason.by.did === requester)
    return true
  return false
}
