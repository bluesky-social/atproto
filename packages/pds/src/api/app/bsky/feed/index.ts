import { Server } from '../../../../lexicon'
import AppContext from '../../../../context'
import getActorFeeds from './getActorFeeds'
import getActorLikes from './getActorLikes'
import getAuthorFeed from './getAuthorFeed'
import getFeed from './getFeed'
import getFeedGenerator from './getFeedGenerator'
import getFeedGenerators from './getFeedGenerators'
import getLikes from './getLikes'
import getListFeed from './getListFeed'
import getPosts from './getPosts'
import getPostThread from './getPostThread'
import getRepostedBy from './getRepostedBy'
import getSuggestedFeeds from './getSuggestedFeeds'
import getTimeline from './getTimeline'
import searchPosts from './searchPosts'

export default function (server: Server, ctx: AppContext) {
  getActorFeeds(server, ctx)
  getActorLikes(server, ctx)
  getAuthorFeed(server, ctx)
  getFeed(server, ctx)
  getFeedGenerator(server, ctx)
  getFeedGenerators(server, ctx)
  getLikes(server, ctx)
  getListFeed(server, ctx)
  getPosts(server, ctx)
  getPostThread(server, ctx)
  getRepostedBy(server, ctx)
  getSuggestedFeeds(server, ctx)
  getTimeline(server, ctx)
  searchPosts(server, ctx)
}
