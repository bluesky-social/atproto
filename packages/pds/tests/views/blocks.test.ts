import assert from 'assert'
import AtpAgent, { AtUri } from '@atproto/api'
import { RecordRef } from '@atproto/bsky/tests/seeds/client'
import { BlockedActorError } from '@atproto/api/src/client/types/app/bsky/feed/getAuthorFeed'
import { BlockedByActorError } from '@atproto/api/src/client/types/app/bsky/feed/getAuthorFeed'
import { isThreadViewPost } from '@atproto/api/src/client/types/app/bsky/feed/defs'
import {
  isViewRecord as isEmbedViewRecord,
  isViewBlocked as isEmbedViewBlocked,
} from '@atproto/api/src/client/types/app/bsky/embed/record'
import { runTestServer, CloseFn, TestServerInfo, forSnapshot } from '../_util'
import { SeedClient } from '../seeds/client'
import basicSeed from '../seeds/basic'

describe('pds views with blocking', () => {
  let server: TestServerInfo
  let agent: AtpAgent
  let close: CloseFn
  let sc: SeedClient
  let danBlockCarol: { uri: string }
  let aliceReplyToDan: { ref: RecordRef }
  let carolReplyToDan: { ref: RecordRef }

  let alice: string
  let carol: string
  let dan: string
  let danBlockUri: string

  beforeAll(async () => {
    server = await runTestServer({
      dbPostgresSchema: 'views_block',
    })
    close = server.close
    agent = new AtpAgent({ service: server.url })
    sc = new SeedClient(agent)
    await basicSeed(sc)
    alice = sc.dids.alice
    carol = sc.dids.carol
    dan = sc.dids.dan
    // add follows to ensure blocks work even w follows
    await sc.follow(carol, dan)
    await sc.follow(dan, carol)
    aliceReplyToDan = await sc.reply(
      alice,
      sc.posts[dan][0].ref,
      sc.posts[dan][0].ref,
      'alice replies to dan',
    )
    carolReplyToDan = await sc.reply(
      carol,
      sc.posts[dan][0].ref,
      sc.posts[dan][0].ref,
      'carol replies to dan',
    )
    // dan blocks carol
    danBlockCarol = await agent.api.app.bsky.graph.block.create(
      { repo: dan },
      { createdAt: new Date().toISOString(), subject: carol },
      sc.getHeaders(dan),
    )
    danBlockUri = danBlockCarol.uri
    await server.processAll()
  })

  afterAll(async () => {
    await close()
  })

  it('blocks thread post', async () => {
    const { data: threadAlice } = await agent.api.app.bsky.feed.getPostThread(
      { depth: 1, uri: sc.posts[carol][0].ref.uriStr },
      { headers: sc.getHeaders(dan) },
    )
    expect(threadAlice).toEqual({
      thread: {
        $type: 'app.bsky.feed.defs#blockedPost',
        uri: sc.posts[carol][0].ref.uriStr,
        blocked: true,
        author: {
          did: carol,
          viewer: {
            blockedBy: false,
            blocking: danBlockUri,
          },
        },
      },
    })
    const { data: threadCarol } = await agent.api.app.bsky.feed.getPostThread(
      { depth: 1, uri: sc.posts[dan][0].ref.uriStr },
      { headers: sc.getHeaders(carol) },
    )
    expect(threadCarol).toEqual({
      thread: {
        $type: 'app.bsky.feed.defs#blockedPost',
        uri: sc.posts[dan][0].ref.uriStr,
        blocked: true,
        author: {
          did: dan,
          viewer: {
            blockedBy: true,
          },
        },
      },
    })
  })

  it('blocks thread reply', async () => {
    // Contains reply by carol
    const { data: thread } = await agent.api.app.bsky.feed.getPostThread(
      { depth: 1, uri: sc.posts[alice][1].ref.uriStr },
      { headers: sc.getHeaders(dan) },
    )
    expect(forSnapshot(thread)).toMatchSnapshot()
  })

  it('blocks thread parent', async () => {
    // Parent is a post by dan
    const { data: thread } = await agent.api.app.bsky.feed.getPostThread(
      { depth: 1, uri: aliceReplyToDan.ref.uriStr },
      { headers: sc.getHeaders(carol) },
    )
    expect(forSnapshot(thread)).toMatchSnapshot()
  })

  it('blocks record embeds', async () => {
    // Contains a deep embed of carol's post, blocked by dan
    const { data: thread } = await agent.api.app.bsky.feed.getPostThread(
      { depth: 0, uri: sc.posts[alice][2].ref.uriStr },
      { headers: sc.getHeaders(dan) },
    )
    expect(forSnapshot(thread)).toMatchSnapshot()
  })

  it('errors on getting author feed', async () => {
    const attempt1 = agent.api.app.bsky.feed.getAuthorFeed(
      { actor: carol },
      { headers: sc.getHeaders(dan) },
    )
    await expect(attempt1).rejects.toThrow(BlockedActorError)

    const attempt2 = agent.api.app.bsky.feed.getAuthorFeed(
      { actor: dan },
      { headers: sc.getHeaders(carol) },
    )
    await expect(attempt2).rejects.toThrow(BlockedByActorError)
  })

  it('strips blocked users out of getTimeline', async () => {
    const resCarol = await agent.api.app.bsky.feed.getTimeline(
      { limit: 100 },
      { headers: sc.getHeaders(carol) },
    )
    expect(
      resCarol.data.feed.some((post) => post.post.author.did === dan),
    ).toBeFalsy()

    const resDan = await agent.api.app.bsky.feed.getTimeline(
      { limit: 100 },
      { headers: sc.getHeaders(dan) },
    )
    expect(
      resDan.data.feed.some((post) => post.post.author.did === carol),
    ).toBeFalsy()
  })

  it('strips blocked users out of getPopular', async () => {
    for (let i = 0; i < 15; i++) {
      const name = `user${i}`
      await sc.createAccount(name, {
        handle: `user${i}.test`,
        email: `user${i}@test.com`,
        password: 'password',
      })
      await sc.like(sc.dids[name], sc.posts[alice][0].ref)
      await sc.like(sc.dids[name], sc.posts[carol][0].ref)
      await sc.like(sc.dids[name], sc.posts[dan][0].ref)
    }

    const resCarol = await agent.api.app.bsky.unspecced.getPopular(
      {},
      { headers: sc.getHeaders(carol) },
    )
    expect(
      resCarol.data.feed.some((post) => post.post.author.did === alice),
    ).toBeTruthy()
    expect(
      resCarol.data.feed.some((post) => post.post.author.did === carol),
    ).toBeTruthy()
    expect(
      resCarol.data.feed.some((post) => post.post.author.did === dan),
    ).toBeFalsy()

    const resDan = await agent.api.app.bsky.unspecced.getPopular(
      {},
      { headers: sc.getHeaders(dan) },
    )
    expect(
      resDan.data.feed.some((post) => post.post.author.did === alice),
    ).toBeTruthy()
    expect(
      resDan.data.feed.some((post) => post.post.author.did === carol),
    ).toBeFalsy()
    expect(
      resDan.data.feed.some((post) => post.post.author.did === dan),
    ).toBeTruthy()
  })

  it('returns block status on getProfile', async () => {
    const resCarol = await agent.api.app.bsky.actor.getProfile(
      { actor: dan },
      { headers: sc.getHeaders(carol) },
    )
    expect(resCarol.data.viewer?.blocking).toBeUndefined
    expect(resCarol.data.viewer?.blockedBy).toBe(true)

    const resDan = await agent.api.app.bsky.actor.getProfile(
      { actor: carol },
      { headers: sc.getHeaders(dan) },
    )
    expect(resDan.data.viewer?.blocking).toBeDefined
    expect(resDan.data.viewer?.blockedBy).toBe(false)
  })

  it('returns block status on getProfiles', async () => {
    const resCarol = await agent.api.app.bsky.actor.getProfiles(
      { actors: [alice, dan] },
      { headers: sc.getHeaders(carol) },
    )
    expect(resCarol.data.profiles[0].viewer?.blocking).toBeUndefined()
    expect(resCarol.data.profiles[0].viewer?.blockedBy).toBe(false)
    expect(resCarol.data.profiles[1].viewer?.blocking).toBeUndefined()
    expect(resCarol.data.profiles[1].viewer?.blockedBy).toBe(true)

    const resDan = await agent.api.app.bsky.actor.getProfiles(
      { actors: [alice, carol] },
      { headers: sc.getHeaders(dan) },
    )
    expect(resDan.data.profiles[0].viewer?.blocking).toBeUndefined()
    expect(resDan.data.profiles[0].viewer?.blockedBy).toBe(false)
    expect(resDan.data.profiles[1].viewer?.blocking).toBeDefined()
    expect(resDan.data.profiles[1].viewer?.blockedBy).toBe(false)
  })

  it('does not return block violating follows', async () => {
    const resCarol = await agent.api.app.bsky.graph.getFollows(
      { actor: carol },
      { headers: sc.getHeaders(alice) },
    )
    expect(resCarol.data.follows.some((f) => f.did === dan)).toBe(false)

    const resDan = await agent.api.app.bsky.graph.getFollows(
      { actor: dan },
      { headers: sc.getHeaders(alice) },
    )
    expect(resDan.data.follows.some((f) => f.did === carol)).toBe(false)
  })

  it('does not return block violating followers', async () => {
    const resCarol = await agent.api.app.bsky.graph.getFollowers(
      { actor: carol },
      { headers: sc.getHeaders(alice) },
    )
    expect(resCarol.data.followers.some((f) => f.did === dan)).toBe(false)

    const resDan = await agent.api.app.bsky.graph.getFollowers(
      { actor: dan },
      { headers: sc.getHeaders(alice) },
    )
    expect(resDan.data.followers.some((f) => f.did === carol)).toBe(false)
  })

  it('does not return posts from blocked users', async () => {
    const alicePost = sc.posts[alice][0].ref.uriStr
    const carolPost = sc.posts[carol][0].ref.uriStr
    const danPost = sc.posts[dan][0].ref.uriStr

    const resCarol = await agent.api.app.bsky.feed.getPosts(
      { uris: [alicePost, carolPost, danPost] },
      { headers: sc.getHeaders(carol) },
    )
    expect(resCarol.data.posts.some((p) => p.uri === alicePost)).toBe(true)
    expect(resCarol.data.posts.some((p) => p.uri === carolPost)).toBe(true)
    expect(resCarol.data.posts.some((p) => p.uri === danPost)).toBe(false)

    const resDan = await agent.api.app.bsky.feed.getPosts(
      { uris: [alicePost, carolPost, danPost] },
      { headers: sc.getHeaders(dan) },
    )
    expect(resDan.data.posts.some((p) => p.uri === alicePost)).toBe(true)
    expect(resDan.data.posts.some((p) => p.uri === carolPost)).toBe(false)
    expect(resDan.data.posts.some((p) => p.uri === danPost)).toBe(true)
  })

  it('does not return notifs for blocked accounts', async () => {
    const resCarol = await agent.api.app.bsky.notification.listNotifications(
      {
        limit: 100,
      },
      { headers: sc.getHeaders(carol) },
    )
    expect(
      resCarol.data.notifications.some((notif) => notif.author.did === dan),
    ).toBeFalsy()

    const resDan = await agent.api.app.bsky.notification.listNotifications(
      {
        limit: 100,
      },
      { headers: sc.getHeaders(dan) },
    )
    expect(
      resDan.data.notifications.some((notif) => notif.author.did === carol),
    ).toBeFalsy()
  })

  it('does not return blocked accounts in actor search', async () => {
    const resCarol = await agent.api.app.bsky.actor.searchActors(
      {
        term: 'dan.test',
      },
      { headers: sc.getHeaders(carol) },
    )
    expect(resCarol.data.actors.some((actor) => actor.did === dan)).toBeFalsy()

    const resDan = await agent.api.app.bsky.actor.searchActors(
      {
        term: 'carol.test',
      },
      { headers: sc.getHeaders(dan) },
    )
    expect(resDan.data.actors.some((actor) => actor.did === carol)).toBeFalsy()
  })

  it('does not return blocked accounts in actor search typeahead', async () => {
    const resCarol = await agent.api.app.bsky.actor.searchActorsTypeahead(
      {
        term: 'dan.test',
      },
      { headers: sc.getHeaders(carol) },
    )
    expect(resCarol.data.actors.some((actor) => actor.did === dan)).toBeFalsy()

    const resDan = await agent.api.app.bsky.actor.searchActorsTypeahead(
      {
        term: 'carol.test',
      },
      { headers: sc.getHeaders(dan) },
    )
    expect(resDan.data.actors.some((actor) => actor.did === carol)).toBeFalsy()
  })

  it('does not return blocked accounts in get suggestions', async () => {
    // unfollow so they _would_ show up in suggestions if not for block
    await sc.unfollow(carol, dan)
    await sc.unfollow(dan, carol)

    const resCarol = await agent.api.app.bsky.actor.getSuggestions(
      {
        limit: 100,
      },
      { headers: sc.getHeaders(carol) },
    )
    expect(resCarol.data.actors.some((actor) => actor.did === dan)).toBeFalsy()

    const resDan = await agent.api.app.bsky.actor.getSuggestions(
      {
        limit: 100,
      },
      { headers: sc.getHeaders(dan) },
    )
    expect(resDan.data.actors.some((actor) => actor.did === carol)).toBeFalsy()
  })

  it('does not serve blocked replies', async () => {
    const getThreadPostUri = (r) => r?.['post']?.['uri']
    // reply then block
    const { data: replyThenBlock } =
      await agent.api.app.bsky.feed.getPostThread(
        { depth: 1, uri: sc.posts[dan][0].ref.uriStr },
        { headers: sc.getHeaders(alice) },
      )
    assert(isThreadViewPost(replyThenBlock.thread))
    expect(replyThenBlock.thread.replies?.map(getThreadPostUri)).toEqual([
      aliceReplyToDan.ref.uriStr,
    ])

    // unblock
    await agent.api.app.bsky.graph.block.delete(
      { repo: dan, rkey: new AtUri(danBlockCarol.uri).rkey },
      sc.getHeaders(dan),
    )
    const { data: unblock } = await agent.api.app.bsky.feed.getPostThread(
      { depth: 1, uri: sc.posts[dan][0].ref.uriStr },
      { headers: sc.getHeaders(alice) },
    )
    assert(isThreadViewPost(unblock.thread))
    expect(unblock.thread.replies?.map(getThreadPostUri)).toEqual([
      carolReplyToDan.ref.uriStr,
      aliceReplyToDan.ref.uriStr,
    ])

    // block then reply
    danBlockCarol = await agent.api.app.bsky.graph.block.create(
      { repo: dan },
      { createdAt: new Date().toISOString(), subject: carol },
      sc.getHeaders(dan),
    )
    const carolReplyToDan2 = await sc.reply(
      carol,
      sc.posts[dan][1].ref,
      sc.posts[dan][1].ref,
      'carol replies to dan again',
    )
    const { data: blockThenReply } =
      await agent.api.app.bsky.feed.getPostThread(
        { depth: 1, uri: sc.posts[dan][0].ref.uriStr },
        { headers: sc.getHeaders(alice) },
      )
    assert(isThreadViewPost(blockThenReply.thread))
    expect(replyThenBlock.thread.replies?.map(getThreadPostUri)).toEqual([
      aliceReplyToDan.ref.uriStr,
    ])

    // cleanup
    await agent.api.app.bsky.feed.post.delete(
      { repo: carol, rkey: carolReplyToDan2.ref.uri.rkey },
      sc.getHeaders(carol),
    )
  })

  it('does not serve blocked embeds to third-party', async () => {
    // embed then block
    const { data: embedThenBlock } =
      await agent.api.app.bsky.feed.getPostThread(
        { depth: 0, uri: sc.posts[dan][1].ref.uriStr },
        { headers: sc.getHeaders(alice) },
      )
    assert(isThreadViewPost(embedThenBlock.thread))
    assert(isEmbedViewBlocked(embedThenBlock.thread.post.embed?.record))

    // unblock
    await agent.api.app.bsky.graph.block.delete(
      { repo: dan, rkey: new AtUri(danBlockCarol.uri).rkey },
      sc.getHeaders(dan),
    )
    const { data: unblock } = await agent.api.app.bsky.feed.getPostThread(
      { depth: 0, uri: sc.posts[dan][1].ref.uriStr },
      { headers: sc.getHeaders(alice) },
    )
    assert(isThreadViewPost(unblock.thread))
    assert(isEmbedViewRecord(unblock.thread.post.embed?.record))

    // block then embed
    danBlockCarol = await agent.api.app.bsky.graph.block.create(
      { repo: dan },
      { createdAt: new Date().toISOString(), subject: carol },
      sc.getHeaders(dan),
    )
    const carolEmbedsDan = await sc.post(
      carol,
      'carol embeds dan',
      undefined,
      undefined,
      sc.posts[dan][0].ref,
    )
    const { data: blockThenEmbed } =
      await agent.api.app.bsky.feed.getPostThread(
        { depth: 0, uri: carolEmbedsDan.ref.uriStr },
        { headers: sc.getHeaders(alice) },
      )
    assert(isThreadViewPost(blockThenEmbed.thread))
    assert(isEmbedViewBlocked(blockThenEmbed.thread.post.embed?.record))

    // cleanup
    await agent.api.app.bsky.feed.post.delete(
      { repo: carol, rkey: carolEmbedsDan.ref.uri.rkey },
      sc.getHeaders(carol),
    )
  })

  it('applies third-party blocking rules in feeds.', async () => {
    // alice follows carol and dan, block exists between carol and dan.
    const replyBlockedUri = carolReplyToDan.ref.uriStr
    const embedBlockedUri = sc.posts[dan][1].ref.uriStr
    const { data: timeline } = await agent.api.app.bsky.feed.getTimeline(
      { limit: 100 },
      { headers: sc.getHeaders(alice) },
    )
    const replyBlockedPost = timeline.feed.find(
      (item) => item.post.uri === replyBlockedUri,
    )
    expect(replyBlockedPost).toBeUndefined()
    const embedBlockedPost = timeline.feed.find(
      (item) => item.post.uri === embedBlockedUri,
    )
    assert(embedBlockedPost)
    assert(isEmbedViewBlocked(embedBlockedPost.post.embed?.record))
  })

  it('returns a list of blocks', async () => {
    await agent.api.app.bsky.graph.block.create(
      { repo: dan },
      { createdAt: new Date().toISOString(), subject: alice },
      sc.getHeaders(dan),
    )

    const res = await agent.api.app.bsky.graph.getBlocks(
      {},
      { headers: sc.getHeaders(dan) },
    )
    const dids = res.data.blocks.map((block) => block.did).sort()
    expect(dids).toEqual([alice, carol].sort())
  })

  it('paginates getBlocks', async () => {
    const full = await agent.api.app.bsky.graph.getBlocks(
      {},
      { headers: sc.getHeaders(dan) },
    )
    const first = await agent.api.app.bsky.graph.getBlocks(
      { limit: 1 },
      { headers: sc.getHeaders(dan) },
    )
    const second = await agent.api.app.bsky.graph.getBlocks(
      { cursor: first.data.cursor },
      { headers: sc.getHeaders(dan) },
    )
    const combined = [...first.data.blocks, ...second.data.blocks]
    expect(combined).toEqual(full.data.blocks)
  })
})
