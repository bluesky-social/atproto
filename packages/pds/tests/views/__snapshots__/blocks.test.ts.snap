// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pds views with blocking blocks record embeds 1`] = `
Object {
  "thread": Object {
    "$type": "app.bsky.feed.defs#threadViewPost",
    "post": Object {
      "author": Object {
        "avatar": "https://pds.public.url/image/KzkHFsMRQ6oAKCHCRKFA1H-rDdc7VOtvEVpUJ82TwyQ/rs:fill:1000:1000:1:0/plain/bafkreiaivizp4xldojmmpuzmiu75cmea7nq56dnntnuhzhsjcb63aou5ei@jpeg",
        "did": "user(0)",
        "displayName": "ali",
        "handle": "alice.test",
        "labels": Array [
          Object {
            "cid": "cids(1)",
            "cts": "1970-01-01T00:00:00.000Z",
            "neg": false,
            "src": "user(0)",
            "uri": "record(2)",
            "val": "self-label-a",
          },
          Object {
            "cid": "cids(1)",
            "cts": "1970-01-01T00:00:00.000Z",
            "neg": false,
            "src": "user(0)",
            "uri": "record(2)",
            "val": "self-label-b",
          },
        ],
        "viewer": Object {
          "blockedBy": false,
          "followedBy": "record(1)",
          "muted": false,
        },
      },
      "cid": "cids(0)",
      "embed": Object {
        "$type": "app.bsky.embed.record#view",
        "record": Object {
          "$type": "app.bsky.embed.record#viewRecord",
          "author": Object {
            "did": "user(1)",
            "handle": "dan.test",
            "labels": Array [
              Object {
                "cts": "1970-01-01T00:00:00.000Z",
                "neg": false,
                "src": "did:example:labeler",
                "uri": "user(1)",
                "val": "repo-action-label",
              },
            ],
            "viewer": Object {
              "blockedBy": false,
              "muted": false,
            },
          },
          "cid": "cids(2)",
          "embeds": Array [
            Object {
              "$type": "app.bsky.embed.record#view",
              "record": Object {
                "$type": "app.bsky.embed.record#viewBlocked",
                "author": Object {
                  "did": "user(2)",
                  "viewer": Object {
                    "blockedBy": false,
                    "blocking": "record(5)",
                  },
                },
                "blocked": true,
                "uri": "record(4)",
              },
            },
          ],
          "indexedAt": "1970-01-01T00:00:00.000Z",
          "labels": Array [],
          "uri": "record(3)",
          "value": Object {
            "$type": "app.bsky.feed.post",
            "createdAt": "1970-01-01T00:00:00.000Z",
            "embed": Object {
              "$type": "app.bsky.embed.record",
              "record": Object {
                "cid": "cids(3)",
                "uri": "record(4)",
              },
            },
            "facets": Array [
              Object {
                "features": Array [
                  Object {
                    "$type": "app.bsky.richtext.facet#mention",
                    "did": "user(0)",
                  },
                ],
                "index": Object {
                  "byteEnd": 18,
                  "byteStart": 0,
                },
              },
            ],
            "text": "@alice.bluesky.xyz is the best",
          },
        },
      },
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "labels": Array [
        Object {
          "cid": "cids(0)",
          "cts": "1970-01-01T00:00:00.000Z",
          "neg": false,
          "src": "did:example:labeler",
          "uri": "record(0)",
          "val": "test-label",
        },
      ],
      "likeCount": 2,
      "record": Object {
        "$type": "app.bsky.feed.post",
        "createdAt": "1970-01-01T00:00:00.000Z",
        "embed": Object {
          "$type": "app.bsky.embed.record",
          "record": Object {
            "cid": "cids(2)",
            "uri": "record(3)",
          },
        },
        "text": "yoohoo label_me",
      },
      "replyCount": 0,
      "repostCount": 0,
      "uri": "record(0)",
      "viewer": Object {},
    },
  },
}
`;

exports[`pds views with blocking blocks thread parent 1`] = `
Object {
  "thread": Object {
    "$type": "app.bsky.feed.defs#threadViewPost",
    "parent": Object {
      "$type": "app.bsky.feed.defs#blockedPost",
      "author": Object {
        "did": "user(1)",
        "viewer": Object {
          "blockedBy": true,
        },
      },
      "blocked": true,
      "uri": "record(4)",
    },
    "post": Object {
      "author": Object {
        "avatar": "https://pds.public.url/image/KzkHFsMRQ6oAKCHCRKFA1H-rDdc7VOtvEVpUJ82TwyQ/rs:fill:1000:1000:1:0/plain/bafkreiaivizp4xldojmmpuzmiu75cmea7nq56dnntnuhzhsjcb63aou5ei@jpeg",
        "did": "user(0)",
        "displayName": "ali",
        "handle": "alice.test",
        "labels": Array [
          Object {
            "cid": "cids(1)",
            "cts": "1970-01-01T00:00:00.000Z",
            "neg": false,
            "src": "user(0)",
            "uri": "record(3)",
            "val": "self-label-a",
          },
          Object {
            "cid": "cids(1)",
            "cts": "1970-01-01T00:00:00.000Z",
            "neg": false,
            "src": "user(0)",
            "uri": "record(3)",
            "val": "self-label-b",
          },
        ],
        "viewer": Object {
          "blockedBy": false,
          "followedBy": "record(2)",
          "following": "record(1)",
          "muted": false,
        },
      },
      "cid": "cids(0)",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "labels": Array [],
      "likeCount": 0,
      "record": Object {
        "$type": "app.bsky.feed.post",
        "createdAt": "1970-01-01T00:00:00.000Z",
        "reply": Object {
          "parent": Object {
            "cid": "cids(2)",
            "uri": "record(4)",
          },
          "root": Object {
            "cid": "cids(2)",
            "uri": "record(4)",
          },
        },
        "text": "alice replies to dan",
      },
      "replyCount": 0,
      "repostCount": 0,
      "uri": "record(0)",
      "viewer": Object {},
    },
    "replies": Array [],
  },
}
`;

exports[`pds views with blocking blocks thread reply 1`] = `
Object {
  "thread": Object {
    "$type": "app.bsky.feed.defs#threadViewPost",
    "post": Object {
      "author": Object {
        "avatar": "https://pds.public.url/image/KzkHFsMRQ6oAKCHCRKFA1H-rDdc7VOtvEVpUJ82TwyQ/rs:fill:1000:1000:1:0/plain/bafkreiaivizp4xldojmmpuzmiu75cmea7nq56dnntnuhzhsjcb63aou5ei@jpeg",
        "did": "user(0)",
        "displayName": "ali",
        "handle": "alice.test",
        "labels": Array [
          Object {
            "cid": "cids(1)",
            "cts": "1970-01-01T00:00:00.000Z",
            "neg": false,
            "src": "user(0)",
            "uri": "record(2)",
            "val": "self-label-a",
          },
          Object {
            "cid": "cids(1)",
            "cts": "1970-01-01T00:00:00.000Z",
            "neg": false,
            "src": "user(0)",
            "uri": "record(2)",
            "val": "self-label-b",
          },
        ],
        "viewer": Object {
          "blockedBy": false,
          "followedBy": "record(1)",
          "muted": false,
        },
      },
      "cid": "cids(0)",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "labels": Array [],
      "likeCount": 3,
      "record": Object {
        "$type": "app.bsky.feed.post",
        "createdAt": "1970-01-01T00:00:00.000000Z",
        "text": "again",
      },
      "replyCount": 2,
      "repostCount": 1,
      "uri": "record(0)",
      "viewer": Object {
        "like": "record(4)",
        "repost": "record(3)",
      },
    },
    "replies": Array [
      Object {
        "$type": "app.bsky.feed.defs#blockedPost",
        "author": Object {
          "did": "user(1)",
          "viewer": Object {
            "blockedBy": false,
            "blocking": "record(6)",
          },
        },
        "blocked": true,
        "uri": "record(5)",
      },
      Object {
        "$type": "app.bsky.feed.defs#threadViewPost",
        "post": Object {
          "author": Object {
            "avatar": "https://pds.public.url/image/KzkHFsMRQ6oAKCHCRKFA1H-rDdc7VOtvEVpUJ82TwyQ/rs:fill:1000:1000:1:0/plain/bafkreiaivizp4xldojmmpuzmiu75cmea7nq56dnntnuhzhsjcb63aou5ei@jpeg",
            "did": "user(2)",
            "displayName": "bobby",
            "handle": "bob.test",
            "labels": Array [
              Object {
                "cid": "cids(3)",
                "cts": "1970-01-01T00:00:00.000Z",
                "neg": false,
                "src": "user(2)",
                "uri": "record(9)",
                "val": "self-label-a",
              },
              Object {
                "cid": "cids(3)",
                "cts": "1970-01-01T00:00:00.000Z",
                "neg": false,
                "src": "user(2)",
                "uri": "record(9)",
                "val": "self-label-b",
              },
            ],
            "viewer": Object {
              "blockedBy": false,
              "following": "record(8)",
              "muted": false,
            },
          },
          "cid": "cids(2)",
          "embed": Object {
            "$type": "app.bsky.embed.images#view",
            "images": Array [
              Object {
                "alt": "tests/image/fixtures/key-landscape-small.jpg",
                "fullsize": "https://pds.public.url/image/AiDXkxVbgBksxb1nfiRn1m6S4K8_mee6o8r-UGLNzOM/rs:fit:2000:2000:1:0/plain/bafkreigy5p3xxceipk2o6nqtnugpft26ol6yleqhboqziino7axvdngtci@jpeg",
                "thumb": "https://pds.public.url/image/uc7FGfiGv0mMqmk9XiqHXrIhNymLHaex7Ge8nEhmXqo/rs:fit:1000:1000:1:0/plain/bafkreigy5p3xxceipk2o6nqtnugpft26ol6yleqhboqziino7axvdngtci@jpeg",
              },
            ],
          },
          "indexedAt": "1970-01-01T00:00:00.000Z",
          "labels": Array [
            Object {
              "cid": "cids(2)",
              "cts": "1970-01-01T00:00:00.000Z",
              "neg": false,
              "src": "did:example:labeler",
              "uri": "record(7)",
              "val": "test-label",
            },
            Object {
              "cid": "cids(2)",
              "cts": "1970-01-01T00:00:00.000Z",
              "neg": false,
              "src": "did:example:labeler",
              "uri": "record(7)",
              "val": "test-label-2",
            },
          ],
          "likeCount": 0,
          "record": Object {
            "$type": "app.bsky.feed.post",
            "createdAt": "1970-01-01T00:00:00.000Z",
            "embed": Object {
              "$type": "app.bsky.embed.images",
              "images": Array [
                Object {
                  "alt": "tests/image/fixtures/key-landscape-small.jpg",
                  "image": Object {
                    "$type": "blob",
                    "mimeType": "image/jpeg",
                    "ref": Object {
                      "$link": "cids(4)",
                    },
                    "size": 4114,
                  },
                },
              ],
            },
            "reply": Object {
              "parent": Object {
                "cid": "cids(0)",
                "uri": "record(0)",
              },
              "root": Object {
                "cid": "cids(0)",
                "uri": "record(0)",
              },
            },
            "text": "hear that label_me label_me_2",
          },
          "replyCount": 1,
          "repostCount": 0,
          "uri": "record(7)",
          "viewer": Object {},
        },
      },
    ],
  },
}
`;
